File: src/main/java/algorithm/Programmers67256.java
package algorithm;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.stream.IntStream;

public class Programmers67256 {
    public String solution(int[] numbers, String hand) {
        StringBuilder answer = new StringBuilder();
        int left = 10, right = 12;
        List<Integer> keypadM = new ArrayList<>(Arrays.asList(2, 5, 8, 11));

        for (int i = 0; i < numbers.length; i++) {
            int n = numbers[i];
            if (n == 0) n = 11;
            switch (n) {
                case 1:
                case 4:
                case 7:
                    left = n;
                    answer.append("L");
                    break;
                case 3:
                case 6:
                case 9:
                    right = n;
                    answer.append("R");
                    break;
                default:
                    int diffL = (keypadM.contains(left)) ? Math.abs(n - left) / 3 : (Math.abs(n - (left + 1)) / 3) + 1;
                    int diffR = (keypadM.contains(right)) ? Math.abs(n - right) / 3 : (Math.abs(n - (right - 1)) / 3) + 1;

                    if (diffL == diffR) {
                        switch (hand) {
                            case "left":
                                left = n;
                                answer.append("L");
                                break;
                            case "right":
                                right = n;
                                answer.append("R");
                                break;
                        }
                    } else if (diffL < diffR) {
                        left = n;
                        answer.append("L");
                    } else {
                        right = n;
                        answer.append("R");
                    }
                    break;
            }
        }
        return answer.toString();
    }

}

File: src/main/java/algorithm/implementation/Programmers67256.kt
package algorithm.implementation

import java.lang.Math.abs

class Programmers67256 {
    val left = "L"
    val right = "R"

    val middles = intArrayOf(2, 5, 8, 0)
    var currentLeft = -1
    var currentRight = -1

    fun solution(numbers: IntArray, hand: String): String {

        var answer = ""

        numbers.forEach {
            answer += when (it) {
                1, 4, 7 -> {
                    setLeft(it)
                }
                3, 6, 9 -> {
                    setRight(it)
                }
                else -> {
                    //2, 5, 8, 0
                    val leftDistance = getDistance(currentLeft, it, 1)
                    val rightDistance = getDistance(currentRight, it, -1)

                    when {
                        leftDistance < rightDistance -> setLeft(it)
                        leftDistance > rightDistance -> setRight(it)
                        else -> if (hand == "left") setLeft(it) else setRight(it)
                    }
                }
            }

        }
        return answer
    }

    private fun getDistance(start: Int, end: Int, makeMiddle: Int): Int {
        return if (middles.contains(start)) {
            abs(middles.indexOf(end) - middles.indexOf(start))
        } else {
            1 + abs(middles.indexOf(end) - middles.indexOf(if (start == -1) 0 else start + makeMiddle))
        }
    }

    private fun setLeft(current: Int): String {
        currentLeft = current
        return left
    }

    private fun setRight(current: Int): String {
        currentRight = current
        return right
    }
}